text/0000-blight-lang.md

```diff
@@ -0,0 +1,159 @@
+language. That being said, I strongly believe that Javascript as a language in
+browsers is going nowhere and will stay the way it is forever.
+
+Without browser support, it's impossible for us to get a good language in
+browsers through traditional means, so until we get a big enough community to
+back us entering browsers by default, we will happily transpile all Blight code
+to Javascript that can run on the server or in the browser.
+
+TLDR; Blight is Typescript 2, inspired by Rust syntax and semantics.
+
+# Motivation
+
+[motivation]: #motivation
+
+We are doing this because we think that with the current landscape of browsers,
+good language features are easy to come by, especially with inspiration from the
+sources. We believe that a language that is designed around scripting with some
+safety checks involved would be amazing and we are committed to making it
+happen.
+
+To be very clear, we do not plan on having any ecosystem to begin. An attack
+strategy is being devised to get around that. We do want to support some sort of
+interface with JavaScript and we plan to do that in a future proposal.
+
+In the best timeline, once we get the compiler up and running, we are able to
+create a runtime pretty fast and kick off a hackathon to get an ecosystem
+running. If we get some key members of the JS community together, we may be able
+to get most of the core necessary modules done pretty fast.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+This proposal is mainly to set up the motivation, goals, and non-goals of the
+language. The motivation has been thoroughly explained in the previous section,
+but let's go over goals and non-goals.
+
+Goals:
+
+- ESM-like modules
+- JS-like async
+- Rust-like syntax
+- Macros
+- Transpiles to JS
+- Fast to compile and produces relatively fast JS
+- Standard library to interact with raw JS APIs safely
+- Zig-like "safe-but-slow" debug build and "fast-but-has-UB" production build
+
+Non-Goals
+
+- Support for any JS libraries outside of wrappers (would just be a worse
+  Typescript)
+- Compile to standalone executable (we'd just be writing Rust again)
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The main consideration for this RFC on the technical side is how will this all
+be implemented. My thinking at the current moment is that we write a
+transpiler + type checker in Rust just for performance sake. I don't think there
+is a universe where self-hosting makes sense unless we somehow get a compiler
+backend.
+
+We could choose other languages as well, but I believe that any other choice is
+the wrong move:
+
+- Zig: Cool technical features, but too unstable to use. Bun suffers actively
+  with breaking changes.
+- C++: It's C++. If I wanted to shoot myself in the foot I wouldn't be in
+  support of Blight.
+- C: Too low level, too much UB, hard to port.
```

Comment:
UB?

Reply:
Undefined behavior. In C if a program invokes undefined behavior at runtime (like signed integer overflow, out-of-bounds array access, or use-after-free), the result of the entire execution is unspecified. So anything can happen.

Reply:
I see, might be worth writing it out, not sure if this is a common acronym most developers know...